group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/TargetAssignmentBuilder.java

```diff
@@ -0,0 +1,327 @@
+                );
+            }
+        });
+
+        // Prepare the topic metadata.
+        Map<Uuid, AssignmentTopicMetadata> topics = new HashMap<>();
+        subscriptionMetadata.forEach((topicName, topicMetadata) ->
+            topics.put(topicMetadata.id(), new AssignmentTopicMetadata(topicMetadata.numPartitions()))
+        );
+
+        // Compute the assignment.
+        GroupAssignment newGroupAssignment = assignor.assign(new AssignmentSpec(
+            Collections.unmodifiableMap(memberSpecs),
+            Collections.unmodifiableMap(topics)
+        ));
+
+        // Compute delta from previous to new assignment and create the
+        // relevant records.
+        List<Record> records = new ArrayList<>();
+        Map<String, Assignment> newTargetAssignment = new HashMap<>();
+
+        memberSpecs.keySet().forEach(memberId -> {
+            Assignment oldMemberAssignment = assignments.get(memberId);
+            Assignment newMemberAssignment = newMemberAssignment(newGroupAssignment, memberId);
+
+            newTargetAssignment.put(memberId, newMemberAssignment);
+
+            if (oldMemberAssignment == null) {
+                // If the member had no assignment, we always create a record for it.
+                records.add(newTargetAssignmentRecord(
+                    groupId,
+                    memberId,
+                    newMemberAssignment.partitions()
+                ));
+            } else {
+                // If the member had an assignment, we only create a record if the
+                // new assignment is different.
+                if (!newMemberAssignment.equals(oldMemberAssignment)) {
+                    records.add(newTargetAssignmentRecord(
+                        groupId,
+                        memberId,
+                        newMemberAssignment.partitions()
+                    ));
+                }
+            }
+        });
+
+        // Bump the assignment epoch.
+        records.add(newTargetAssignmentEpochRecord(groupId, groupEpoch));
+
+        return new TargetAssignmentResult(records, newTargetAssignment);
+    }
+
+    private Assignment newMemberAssignment(
+        GroupAssignment newGroupAssignment,
+        String memberId
+    ) {
+        MemberAssignment newMemberAssignment = newGroupAssignment.members().get(memberId);
+        if (newMemberAssignment != null) {
+            return new Assignment(newMemberAssignment.targetPartitions());
+        } else {
+            return Assignment.EMPTY;
+        }
+    }
+
+    private void addMemberSpec(
+        Map<String, AssignmentMemberSpec> members,
+        ConsumerGroupMember member,
+        Assignment targetAssignment
+    ) {
+        Set<Uuid> subscribedTopics = new HashSet<>();
+        member.subscribedTopicNames().forEach(topicName -> {
+            TopicMetadata topicMetadata = subscriptionMetadata.get(topicName);
+            if (topicMetadata != null) {
+                subscribedTopics.add(topicMetadata.id());
```

Comment:
To confirm -- we always expect to have an ID if the metadata is present.
What about the case where the metadata is not present? We just don't subscribe to the topic?

Reply:
Right. The id should be there because the new group coordinator will only work with kraft. If the metadata is not present, we just don't subscribe to the topic.

Reply:
Do we expect the metadata to be missing if we have the name in subscribedTopicNames? Do we expect to eventually get the metadata?

Reply:
They are two different things. As the consumer subscribe to a name, it could be that the topic does not exist yet for instance. In this case, we just ignore it and revise when the topic is created (if it ever is).